pkg/ingester/circuitbreaker.go

```diff
@@ -195,47 +203,54 @@ func (cb *circuitBreaker) tryRecordFailure(err error) bool {
 }
 
 func (cb *circuitBreaker) isActive() bool {
-	return cb != nil && cb.active.Load()
+	return cb != nil && (cb.state.Load() == circuitBreakerActive)
 }
 
-func (cb *circuitBreaker) isActivationIdle() bool {
-	return cb != nil && (cb.activating.Load() == activationIdle)
-}
-
-func (cb *circuitBreaker) isActivationPending() bool {
-	return cb != nil && (cb.activating.Load() == activationPending)
-}
+func (cb *circuitBreaker) activate() {
+	if cb == nil {
+		return
+	}
 
-func (cb *circuitBreaker) isActivationQueued() bool {
-	return cb != nil && (cb.activating.Load() == activationQueued)
-}
+	cb.stateMtx.Lock()
+	defer cb.stateMtx.Unlock()
 
-func (cb *circuitBreaker) activate() {
-	if cb == nil || cb.active.Load() {
+	if cb.state.Load() != circuitBreakerInactive {
 		return
 	}
+
 	if cb.cfg.InitialDelay > 0 {
-		cb.activating.CompareAndSwap(activationIdle, activationPending)
+		// Move breaker from inactive to pending state. Activation will be completed by tryAcquirePermit()
+		cb.activationNeeded.Store(true)
+		cb.state.Store(circuitBreakerPending)
 	} else {
+		// Immediately move the breaker to active state
 		level.Info(cb.logger).Log("msg", "activating circuit breaker", "requestType", cb.requestType)
-		cb.active.Store(true)
+		cb.state.Store(circuitBreakerActive)
 	}
 }
 
 func (cb *circuitBreaker) deactivate() {
 	if cb == nil {
 		return
 	}
-	if !cb.active.Load() {
-		// cancel delayed activation if it was scheduled
-		if old := cb.activating.Swap(activationIdle); old != activationIdle {
-			level.Info(cb.logger).Log("msg", "canceled delayed circuit breaker activation", "requestType", cb.requestType)
-		}
+
+	cb.stateMtx.Lock()
+	defer cb.stateMtx.Unlock()
+
+	if cb.state.Load() == circuitBreakerInactive {
 		return
 	}
+
 	level.Info(cb.logger).Log("msg", "deactivating circuit breaker", "requestType", cb.requestType)
-	cb.activating.Store(activationIdle)
-	cb.active.Store(false)
+	cb.activationNeeded.Store(false)
+	cb.state.Store(circuitBreakerInactive)
+
+	// cancel any pending activation
+	if cb.activationTimer != nil {
+		cb.activationTimer.Stop()
```

Comment:
I guess here we rely on the fact that within `time.AfterFunc()` we actually set `cb.activationTimer` to `nil`. 
Are we sure it can never happen that we reach this line while `time.AfterFunc()` is executing?

I personally prefer usage of `time.NewTimer()` because then `time.Stop()`'s result (`true` or `false`) clearly explains what happened. Here `false` means that the timer expired, and the function passed to `time.AfterFun()` was started in a separate go routine, but doesn't say anything about whether the function was completed or not. 

Anyway, if we opt for this approach, since it is crucial to execute `cb.activationTimer = nil`, I would put a comment on both lines that execute it (251 and 284) to highlight that those lines must not be deleted or moved from those places, because that would break the logic.